File: problems/108-convert-sorted-array-to-binary-search-tree/solution.c
#include <stdio.h>

struct TreeNode {
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
};

struct TreeNode* sortedArrayToBST(int* nums, int size) {
    if (size == 0) return NULL;

    struct TreeNode* tree = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    int mid = size / 2;

    tree->val = nums[mid];
    tree->left = sortedArrayToBST(nums, mid);
    tree->right = sortedArrayToBST(nums + mid + 1, size - mid - 1);
    return tree;
}

File: problems/108-convert-sorted-array-to-binary-search-tree/solution.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val: int, left: 'TreeNode', right: 'TreeNode'):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:        
        tree = self.rec(nums)
        return tree


    def rec(self, nums: list[int]):
        if not len(nums):
            return None

        mid = len(nums) // 2
        tree = TreeNode(nums[mid])
        tree.left = self.rec(nums[:mid])
        tree.right = self.rec(nums[mid + 1:])

        return tree